### PR TITLE
Add ListFilter component

### PR DESCRIPTION
**Pull request by Boerescu-Razvan**
Port ListFilter component to typescript

### CODE REVIEW

**Comment by alexandra-c** [comment]
Increase the number of tests to cover at least 80% code coverage. 
image

File: src/components/inputs/ListFilter/UserPreferencesPopUp.tsx [hunk 1]
```diff
@@ -0,0 +1,160 @@
+import React, { useCallback, useState } from 'react'
+import PropTypes from 'prop-types'
+import {
```
Comment (alexandra-c) [inline src/components/inputs/ListFilter/UserPreferencesPopUp.tsx:3]: Use the locally defined components not directly the ones from material-ui

File: src/stories/inputs/ListFilter/ListFilterDefaultDecorator.tsx [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  const onAddUserPreference = (popUpUserPreference: UserPreference) => {
+    if (popUpUserPreference.filterName === '') return
+
+    const newPreference = Object.assign({}, popUpUserPreference, {
+      isDefault: false,
+      implicit: false,
+      filterName: popUpUserPreference.filterName.trim()
+    })
```
Comment (alexandra-c) [inline src/stories/inputs/ListFilter/ListFilterDefaultDecorator.tsx:68]: Do not use `Object.assign`, starting from ES6 or so we use the spread operator.
```javascript
const newPreference = {
      ...popUpUserPreference, 
      isDefault: false,
      implicit: false,
      filterName: popUpUserPreference.filterName.trim()
    }
```

File: src/stories/inputs/ListFilter/ListFilterDefaultDecorator.tsx [hunk 3]
```diff
@@ -0,0 +1,190 @@
+import React, { useCallback, useState } from 'react'
```
Comment (alexandra-c) [inline src/stories/inputs/ListFilter/ListFilterDefaultDecorator.tsx:1]: try to use ramda functions instead on lodash

File: src/stories/inputs/ListFilter/ListFilterDefaultDecorator.tsx [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    const index = args.visibleFilters.indexOf(filter)
+    const newVisibleFilters = Object.assign([...args.visibleFilters], { [index]: { ...filter, isVisible: value } })
+    updateArgs({visibleFilters: newVisibleFilters})
+  }
+
+  const filtersToShow = args.visibleFilters.filter((x: VisibleFilter) => x.isVisible).map((x: VisibleFilter) => x.fieldName)
+
+  const handleVisibleFieldsChange = (name: string) => (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = event.target.checked
+    const filter = args.visibleFields.find((x: VisibleField) => x.fieldName === name)
```
Comment (alexandra-c) [inline src/stories/inputs/ListFilter/ListFilterDefaultDecorator.tsx:45]: `visibleFields` can be destructured from args. Simplify your code

**Comment by alexandra-c** [comment]
ListFilter and UserPreferencesPopup files are huge, they probably have low rendering performance. You need to split them in smaller components to reduce the quantity of UI components that will re-render when something changes.
In addition, the code is hard to read and debug being so big.

**Comment by alexandra-c** [comment]
Do not import icons from material-ui directly from the index file. Use specific imports. material-ui does not have tree-shaking so it will include all the icons in our bundle.

Spcific imports:
```javascript
import Close from '@mui/icons-material/Close'
```

File: src/components/inputs/ListFilter/ListFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,530 @@
+import React, { useCallback, useMemo, useState } from 'react'
+import PropTypes from 'prop-types'
+import { Toolbar, Tooltip, Collapse, Menu, MenuItem, FormControlLabel, Checkbox } from '@mui/material/'
+import { Grid } from '@mui/material'
+import TextField from '../TextField'
+import InputAdornment from '@mui/material/InputAdornment'
+import IconButton from '@mui/material/IconButton'
```
Comment (alexandra-c) [inline src/components/inputs/ListFilter/ListFilter.tsx:7]: Use local components, from our library, not the ones from material-UI

File: src/components/inputs/ListFilter/ListFilter.tsx [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+                  <div>
+                    <CustomIconButton color='primary' onClick={openVisibleFieldsMenu}>
+                      <FilterList />
+                    </CustomIconButton>
+                  </div>
+                </Tooltip>
+              </Grid>
+            )}
+          </Grid>
+          {/* VisibleFiltersMenu */}
```
Comment (alexandra-c) [inline src/components/inputs/ListFilter/ListFilter.tsx:311]: you forgot to delete this comment?

File: src/stories/inputs/ListFilter/_hooks.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import { useEffect, useState } from 'react'
+import { listData } from './_mocks'
+import { Filters } from 'components'
+
+export function useFilteredList(filters: Filters) {
+  const [localListData, setLocalListData] = useState(listData)
+  
+  useEffect(()=>{
+    let newListData = [...listData]
```
[the remark this answers is not in the document]
Comment (st-angelo) [reply src/stories/inputs/ListFilter/_hooks.ts:9]: I may be wrong, but doesn't directly mutating an imported variable (with .sort() for instance) change its value for further uses? Like if further in the function (or in other functions or even a different file) we need to get the mock data again, its value will have changed and in this case that should never happen.
Maybe `structuredClone` should be used here?

[the remark this answers is not in the document]
Comment (alexandra-c) [reply src/stories/inputs/ListFilter/_hooks.ts:9]: If you use Ramda, as i suggested in the next comments, by using `let newListData = listData` any changes made to the `newListData` will only apply to the `newListData`, the `listData` variable will remain unchanged.
Ramda functions are automatically immutable.
`Ramda emphasizes a purer functional style. Immutability and side-effect free functions are at the heart of its design philosophy. `
You can try this if you want:
https://ramdajs.com/repl/#?const%20pi%20%3D%20%5B4.122222222%2C%204.144444%5D%0Alet%20pi2%3Dpi%0Api2%20%3D%20R.map%28v%3D%3Ev%2B1%2C%20pi2%29%0A%0Api%20%2F%2F%20%5B4.122222222%2C%204.144444%5D%0Api2%20%2F%2F%20%5B5.122222222%2C%205.144444%2C%205.1%5D%0A%0A
image

File: src/components/inputs/ListFilter/FullTextFilterEndAdornment.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import InputAdornment from '@mui/material/InputAdornment'
+import { FullTextFilterEndAdornmentProps } from './types'
+import { Close, ExpandLess, ExpandMore, FilterList } from '@mui/icons-material'
```
Comment (DCosti) [inline src/components/inputs/ListFilter/FullTextFilterEndAdornment.tsx:5]: Importa fiecare icon separat deoarece `mui/icons-material` nu are three-shaking. Verifica te rog peste tot si schimba unde e nevoie.

File: src/components/inputs/ListFilter/FullTextFilterEndAdornment.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import { IconButton } from 'components'
+
+const FullTextFilterEndAdornment = ({ localizedStrings, resetTextFilter, expandFilters, expanded, openVisibleFiltersMenu, hasChildren, hasVisibleFilters } : FullTextFilterEndAdornmentProps) => {
+  return (
+    <InputAdornment position="end">
+    <IconButton size='small' color='transparent' aria-label='Reset Filters' aria-description={localizedStrings.ResetFilters} tooltip={localizedStrings.ResetFilters} onClick={resetTextFilter}>
```
Comment (DCosti) [inline src/components/inputs/ListFilter/FullTextFilterEndAdornment.tsx:11]: IconButton-ul facut de noi are o proprietate `type` din care poti alege mai multe variante: 
image
Incearca sa folosesti IconButton-ul local cu unul dina ceste type-uri acolo udne este posibil.

File: src/components/inputs/ListFilter/ListFilter.test.tsx [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+    dbColumnName: 'name',
+    displayName: 'Name'
+  },
+  {
+    dbColumnName: 'code',
+    displayName: 'Code'
+  }
+]
+
+export const visibleFilters: ListFilterProps['visibleFilters'] = [
```
Comment (DCosti) [inline src/components/inputs/ListFilter/ListFilter.test.tsx:49]: unde mai folosesti aceste const-uri de le exporti de aici?

File: src/components/inputs/ListFilter/ListFilter.tsx [hunk 11]
```diff
@@ -0,0 +1,473 @@
+import React, { useCallback, useMemo, useState } from 'react'
+import PropTypes from 'prop-types'
+import { TextField, Autocomplete } from 'components'
+import { Toolbar, Collapse, Menu, MenuItem, InputAdornment, Grid } from '@mui/material/'
```
Comment (DCosti) [inline src/components/inputs/ListFilter/ListFilter.tsx:4]: ai un `/` la final

File: src/components/inputs/ListFilter/types.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
+import { ReactNode } from 'react'
+
+export interface SortableColumn {
+  dbColumnName: string
```
Comment (DCosti) [inline src/components/inputs/ListFilter/types.ts:4]: Nu se poate numi `columnName`?

File: src/components/inputs/ListFilter/FullTextFilterEndAdornment.tsx (hunk 9 above)
Comment (Boerescu-Razvan) [reply src/components/inputs/ListFilter/FullTextFilterEndAdornment.tsx:14]: am avut doar un caz unde am putut folosi varianta de download

File: src/components/inputs/ListFilter/UserPreferencesPopUp.tsx [hunk 13]
```diff
@@ -0,0 +1,87 @@
+import React from 'react'
```
Comment (DCosti) [inline src/components/inputs/ListFilter/UserPreferencesPopUp.tsx:1]: Nu poti folosi componenta de Dialog pe care o avem ca sa nu mai fie nevoie de aceasta componenta?

File: src/components/inputs/ListFilter/FullTextFilterEndAdornment.tsx [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+import ExpandLessIcon from '@mui/icons-material/ExpandLess'
+import ExpandMoreIcon from '@mui/icons-material/ExpandMore'
+import FilterListIcon from '@mui/icons-material/FilterList'
+import { IconButton } from 'components'
+
+const FullTextFilterEndAdornment = ({ localizedStrings, resetTextFilter, expandFilters, expanded, openVisibleFiltersMenu, hasChildren, hasVisibleFilters } : FullTextFilterEndAdornmentProps) => {
+  return (
+    <InputAdornment position="end">
+    <IconButton size='small' color='transparent' aria-label='Reset Filters' aria-description={localizedStrings.ResetFilters} tooltip={localizedStrings.ResetFilters} onClick={resetTextFilter}>
+      <CloseIcon />
```
Comment (DCosti) [inline src/components/inputs/ListFilter/FullTextFilterEndAdornment.tsx:15]: avem deja un `type:close` in IconButton.De ce a fost nevoie sa pui tu icon-ul de Close?

Comment (Boerescu-Razvan) [reply src/components/inputs/ListFilter/FullTextFilterEndAdornment.tsx]: Eu n-am vazut sa existe `type:close`, am vazut in schimb `type:cancel`, dar nu mi se parea ca arata la fel de bine. In orice caz, am adaugat acum type-uri in IconButton pentru toate icon-urile pe care le-am folosit

File: src/components/buttons/IconButton/types.ts [hunk 15]
```diff
@@ -4,28 +4,38 @@
...
 import ArrowUpwardIcon from '@mui/icons-material/ArrowUpward'
+import ExpandLessIcon from '@mui/icons-material/ExpandLess'
+import ExpandMoreIcon from '@mui/icons-material/ExpandMore'
+import FilterListIcon from '@mui/icons-material/FilterList'
+import PostAddIcon from '@mui/icons-material/PostAdd'
 
 export type FontSize = 'inherit' | 'small' | 'medium' | 'large'
 
-export type ButtonType = 'add' | 'cancel' | 'delete' | 'download' | 'downward' | 'edit' | 'view' | 'save' | 'upward'
+export type ButtonType = 'add' | 'cancel' | 'close' | 'delete' | 'download' | 'downward' | 'edit' | 'view' | 'save' | 'upward' | 'expandLess' | 'expandMore' | 'filterList' | 'postAdd'
```
Comment (DCosti) [inline src/components/buttons/IconButton/types.ts:22]: 1. `filterList` si `postAdd` sunt prea specifice pentru a fi adaugate in lista de `type` din IconButton.
2. Scoate `close` si inlocuieste icon-ul de la `cancel` cu ce ai pus la `close`.
3. Exista un story ptr IconButton, Types, in care trebuie adaugate cele doua type-uri noi, `expandMore`, `expandLess` https://osstotalsoft.github.io/rocket-ui-ts/?path=/docs/components-buttons-iconbutton--docs#types

**Review by alexandra-c — APPROVED** [review]

**Pull request merged**